### PR TITLE
ftrace 

### PR DESCRIPTION
**Pull request by briankoco**
Introduce basic ftrace functionality
 - This relies on the "-pg" flag to build the kernel with tracing. All kernel functions are interposed with a call to mcount(), which is responsible for saving the running register context and collecting information about the function being invoked and storing in a ringbuffer.
 - Currently, a user process has to set itself to be traced and then read the /sys/ files to collect tracing information. A future PR will provide external tracing abilities

### CODE REVIEW

File: src/unix/ftrace.c [hunk 1]
```diff
@@ -0,0 +1,1103 @@
+#include <runtime.h>
+#include <unix_internal.h>
+#include <ftrace.h>
+#include <x86_64.h>
+#include <symtab.h>
+
+#ifdef CONFIG_FTRACE
```
Comment (wjhun) [inline src/unix/ftrace.c]: Maybe conditional compile rather than #ifdef for the entire file?

File: src/unix/ftrace.c [hunk 2]
```diff
@@ -0,0 +1,1103 @@
...
+    }
+
+    rv = do_trace_pipe_read(
+        bound(p), rbuf, bound(buf), bound(length), bound(offset)
+    );
+
+    if (rv == 0) {
+        if (!blocked)
+            rbuf_wait(rbuf);
+        rv = infinity;
```
Comment (wjhun) [inline src/unix/ftrace.c]: Will need to change this to BLOCKQ_BLOCK_REQUIRED

File: src/unix/ftrace.c [hunk 3]
```diff
@@ -0,0 +1,1103 @@
...
+
+    unsigned long disable_cnt;
+    unsigned long wait_cnt;
+    timer trace_timer;
+    blockq bq;
+};
+
+/* This structure is designed to simply the process of efficiently
+ * writing buffers to userspace when the user issues multiple reads
+ * with changing offsets
```
Comment (wjhun) [inline src/unix/ftrace.c]: I think we can sort of safely assume that there won't be any reading of this ftrace data from userspace - unless we want to consider the case of having a diagnostic user program that happens to inspect ftrace data? Seems doubtful for the short-term anyway.

More likely is that buffers of data will be shipped to be sent out remotely, as in the case of the http and/or direct socket interface. In this case, said buffers will sit on a queue waiting to transmitted, and offsets within are tracked using buffer_consume (i.e. b->start). Of course, this is destructive, so any read of persistent data will need to do a clone_buffer or something similar. However this is all fine to merge as-is; we can make adjustments as we wire up the remote interface(s). And we may still want that kernel diagnostic userland test down the road (e.g. a performance / memory regression test).

File: src/x86_64/ftrace.s [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+;;  [ tracee retaddr in parent ]  <--- moved into rsi 
+%macro save_mcount_regs 0
+    push rbp
+    push qword [rsp+8]
+    push rbp
+    mov rbp, rsp
+ 
+    ;; we add enough stack to save all the regs, but only need those
+    ;; needed and potentially clobbered by mcount
+    sub rsp, (SAVE_REG_SIZE - SAVE_FRAME_SIZE)
```
Comment (wjhun) [inline src/x86_64/ftrace.s]: I haven't worked through all this, but can we be assured that rsp always stays 16-byte aligned?

File: stage3/Makefile [hunk 5]
```diff
@@ -40,6 +40,7 @@ SRCS-stage3.img= \
 	$(SRCDIR)/unix/blockq.c \
 	$(SRCDIR)/unix/exec.c \
 	$(SRCDIR)/unix/eventfd.c \
+	$(SRCDIR)/unix/ftrace.c \
```
Comment (wjhun) [inline stage3/Makefile]: Again maybe make these conditional on CONFIG_FTRACE?

File: src/unix/ftrace.c (hunk 3 above)
Comment (briankoco) [reply src/unix/ftrace.c]: Yeah, this was all done before realizing that we didn't really need a userland interface. I do think the ftrace_printer stuff might be worth separating into a standalone interface at some point, if we have an application that needs to read large virtual files. So I'll leave it as is for the time being.

File: src/x86_64/ftrace.s (hunk 4 above)
Comment (briankoco) [reply src/x86_64/ftrace.s]: good call, thanks

**Pull request merged**